File: frontend/src/api/actions.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { apiClient } from "./client";
import { toast } from "../store/toastStore";

export interface FetchEmailsLogEntry {
  time: string;
  message: string;
}

export interface FetchEmailsStatus {
  running: boolean;
  phase: "idle" | "listing" | "fetching" | "done" | "cancelled" | "error";
  done: number;
  total: number;
  fetched: number;
  error: string | null;
  logs: FetchEmailsLogEntry[];
}

export function useFetchEmails() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async () => {
      const { data } = await apiClient.post<{ status: string }>("/actions/fetch-emails");
      return data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["emails"] });
      toast.info("Fetching emails from Gmail…");
    },
  });
}

export function useCancelFetchEmails() {
  return useMutation({
    mutationFn: async () => {
      const { data } = await apiClient.post<{ status: string }>("/actions/fetch-emails/cancel");
      return data;
    },
    onSuccess: () => toast.info("Cancelling fetch…"),
  });
}

export function useFetchEmailsStatus() {
  return useQuery({
    queryKey: ["actions", "fetch-emails-status"],
    queryFn: async () => {
      const { data } = await apiClient.get<FetchEmailsStatus>("/actions/fetch-emails/status");
      return data;
    },
    // Always poll on mount so a page reload re-attaches to whatever the
    // server is actually doing, instead of trusting client-only mutation state.
    refetchInterval: (query) => (query.state.data?.running ? 1500 : false),
  });
}

export interface ProcessPendingStatus {
  running: boolean;
  processed: number;
  failed: number;
  error: string | null;
}

export function useProcessPending() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async () => {
      const { data } = await apiClient.post<{ status: string }>("/actions/process-pending");
      return data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["emails"] });
      toast.info("Processing pending emails into offers…");
    },
  });
}

export function useProcessPendingStatus(enabled: boolean) {
  return useQuery({
    queryKey: ["actions", "process-pending-status"],
    queryFn: async () => {
      const { data } = await apiClient.get<ProcessPendingStatus>("/actions/process-pending/status");
      return data;
    },
    enabled,
    refetchInterval: enabled ? 1500 : false,
  });
}

export interface FetchSalesWebStatus {
  running: boolean;
  saved: number;
  failed: number;
  total_active: number;
  done: number;
  total: number;
}

export function useFetchSalesWeb() {
  return useMutation({
    mutationFn: async () => {
      const { data } = await apiClient.post<{ status: string }>("/actions/fetch-sales-web");
      return data;
    },
    onSuccess: () => toast.info("Searching the web for sales — this may take a while."),
  });
}

export function useFetchSalesWebStatus(enabled: boolean) {
  return useQuery({
    queryKey: ["actions", "fetch-sales-web-status"],
    queryFn: async () => {
      const { data } = await apiClient.get<FetchSalesWebStatus>("/actions/fetch-sales-web/status");
      return data;
    },
    enabled,
    refetchInterval: enabled ? 1500 : false,
  });
}

export interface ResearchBrandsStatus {
  running: boolean;
  done: number;
  total: number;
  inserted: number;
  updated: number;
  failed_brands: string[];
}

export function useResearchBrands() {
  return useMutation({
    mutationFn: async () => {
      const { data } = await apiClient.post<{ status: string }>("/actions/research-brands");
      return data;
    },
    onSuccess: () => toast.info("Researching brands via Tavily + Llama — this may take a while."),
  });
}

export function useResearchBrandsStatus(enabled: boolean) {
  return useQuery({
    queryKey: ["actions", "research-brands-status"],
    queryFn: async () => {
      const { data } = await apiClient.get<ResearchBrandsStatus>("/actions/research-brands/status");
      return data;
    },
    enabled,
    refetchInterval: enabled ? 1500 : false,
  });
}

File: api/actions.py
"""One-off pipeline action endpoints — mirrors the Streamlit sidebar action buttons."""
import threading
from datetime import datetime, timezone

from flask import Blueprint, jsonify

bp = Blueprint("actions", __name__, url_prefix="/api/actions")

_web_sales_state = {"running": False, "saved": 0, "failed": 0, "total_active": 0, "done": 0, "total": 0}
_research_state = {"running": False, "done": 0, "total": 0, "inserted": 0, "updated": 0, "failed_brands": []}
_process_pending_state = {"running": False, "processed": 0, "failed": 0, "error": None}

_FETCH_LOG_LIMIT = 40
_fetch_emails_state = {
    "running": False,
    "phase": "idle",  # idle | listing | fetching | done | cancelled | error
    "done": 0,
    "total": 0,
    "fetched": 0,
    "error": None,
    "logs": [],
}
_fetch_cancel_event = threading.Event()


def _fetch_log(message: str) -> None:
    _fetch_emails_state["logs"].append({
        "time": datetime.now(timezone.utc).replace(tzinfo=None).isoformat(),
        "message": message,
    })
    del _fetch_emails_state["logs"][:-_FETCH_LOG_LIMIT]


def _fetch_progress(event: str, data: dict) -> None:
    if event == "listing":
        _fetch_emails_state["phase"] = "listing"
        _fetch_log("Listing messages under the Gmail label…")
    elif event == "found":
        _fetch_emails_state["phase"] = "fetching"
        _fetch_emails_state["total"] = data["total"]
        _fetch_log(f"Found {data['total']} new message(s) to fetch.")
    elif event == "message":
        _fetch_emails_state["done"] = data["done"]
        if data.get("error"):
            _fetch_log(f"[{data['done']}/{data['total']}] failed: {data['error']}")
        else:
            _fetch_emails_state["fetched"] += 1
            subject = (data.get("subject") or "")[:70]
            _fetch_log(f"[{data['done']}/{data['total']}] fetched — {subject}")
    elif event == "cancelled":
        _fetch_emails_state["phase"] = "cancelled"
        _fetch_log(f"Cancelled after {data['done']}/{data['total']} message(s).")


def _run_fetch_emails():
    from gmail.gmail_service import fetch_and_store_emails

    try:
        saved = fetch_and_store_emails(on_progress=_fetch_progress, cancel_event=_fetch_cancel_event)
        _fetch_emails_state["fetched"] = len(saved)
        if _fetch_emails_state["phase"] != "cancelled":
            _fetch_emails_state["phase"] = "done"
            _fetch_log(f"Done — {len(saved)} new email(s) saved.")
    except Exception as exc:
        _fetch_emails_state["phase"] = "error"
        _fetch_emails_state["error"] = str(exc)
        _fetch_log(f"Error: {exc}")
    finally:
        _fetch_emails_state["running"] = False


@bp.post("/fetch-emails")
def fetch_emails():
    if _fetch_emails_state["running"]:
        return jsonify({"status": "already_running"}), 409

    _fetch_cancel_event.clear()
    _fetch_emails_state.update({
        "running": True, "phase": "listing", "done": 0, "total": 0, "fetched": 0, "error": None, "logs": [],
    })
    threading.Thread(target=_run_fetch_emails, daemon=True).start()
    return jsonify({"status": "started"}), 202


@bp.get("/fetch-emails/status")
def fetch_emails_status():
    return jsonify(_fetch_emails_state)


@bp.post("/fetch-emails/cancel")
def fetch_emails_cancel():
    if not _fetch_emails_state["running"]:
        return jsonify({"status": "not_running"}), 409
    _fetch_cancel_event.set()
    _fetch_log("Cancellation requested…")
    return jsonify({"status": "cancelling"})


def _run_process_pending():
    from scheduler.jobs import process_emails_without_offers

    try:
        processed, failed = process_emails_without_offers()
        _process_pending_state.update({"processed": processed, "failed": failed, "error": None})
    except Exception as exc:
        _process_pending_state["error"] = str(exc)
    finally:
        _process_pending_state["running"] = False


@bp.post("/process-pending")
def process_pending():
    if _process_pending_state["running"]:
        return jsonify({"status": "already_running"}), 409

    _process_pending_state.update({"running": True, "processed": 0, "failed": 0, "error": None})
    threading.Thread(target=_run_process_pending, daemon=True).start()
    return jsonify({"status": "started"}), 202


@bp.get("/process-pending/status")
def process_pending_status():
    return jsonify(_process_pending_state)


def _run_fetch_sales_web():
    from ai import cache as _ai_cache
    from ai.brand_fetcher import fetch_offers_for_brand, load_brands
    from database.db import get_session
    from database.models import Offer

    brands = load_brands()
    _web_sales_state.update({
        "running": True, "saved": 0, "failed": 0, "total_active": 0, "done": 0, "total": len(brands),
    })

    cache = _ai_cache.load()
    for brand in brands:
        try:
            offers_data = fetch_offers_for_brand(brand, cache=cache)
            for o_data in offers_data:
                try:
                    offer = Offer(
                        email_id=None,
                        brand=o_data.get("brand"),
                        company=o_data.get("company"),
                        category=o_data.get("category"),
                        subcategory=o_data.get("subcategory"),
                        offer_type=o_data.get("offer_type"),
                        discount_percentage=o_data.get("discount_percentage"),
                        coupon_code=o_data.get("coupon_code"),
                        expiry_date=o_data.get("expiry_date"),
                        offer_value=o_data.get("offer_value"),
                        website=o_data.get("website") or o_data.get("_url") or None,
                        summary=o_data.get("summary"),
                        key_highlights=o_data.get("key_highlights"),
                        is_active=bool(o_data.get("is_active", False)),
                        source="ai",
                    )
                    with get_session() as session:
                        session.add(offer)
                    _web_sales_state["saved"] += 1
                    if o_data.get("_is_sale"):
                        _web_sales_state["total_active"] += 1
                except Exception:
                    _web_sales_state["failed"] += 1
        except Exception:
            _web_sales_state["failed"] += 1
        finally:
            _web_sales_state["done"] += 1

    _ai_cache.save(cache)
    _web_sales_state["running"] = False


@bp.post("/fetch-sales-web")
def fetch_sales_web():
    from ai.brand_fetcher import load_brands

    if _web_sales_state["running"]:
        return jsonify({"status": "already_running"}), 409

    if not load_brands():
        return jsonify({"error": "brands.json is empty or missing"}), 400

    threading.Thread(target=_run_fetch_sales_web, daemon=True).start()
    return jsonify({"status": "started"}), 202


@bp.get("/fetch-sales-web/status")
def fetch_sales_web_status():
    return jsonify(_web_sales_state)


def _run_research_brands():
    from research.runner import ResearchRunner

    runner = ResearchRunner()
    active_brands = runner._db.get_active_brands()
    _research_state.update({
        "running": True, "done": 0, "total": len(active_brands), "inserted": 0, "updated": 0, "failed_brands": [],
    })

    results = runner.run()
    _research_state["inserted"] = sum(r.inserted for r in results)
    _research_state["updated"] = sum(r.updated for r in results)
    _research_state["failed_brands"] = [r.brand_name for r in results if not r.success]
    _research_state["done"] = len(results)
    _research_state["running"] = False


@bp.post("/research-brands")
def research_brands():
    from research.config import TAVILY_API_KEY

    if not TAVILY_API_KEY:
        return jsonify({"error": "TAVILY_API_KEY is not set in .env"}), 400
    if _research_state["running"]:
        return jsonify({"status": "already_running"}), 409

    threading.Thread(target=_run_research_brands, daemon=True).start()
    return jsonify({"status": "started"}), 202


@bp.get("/research-brands/status")
def research_brands_status():
    return jsonify(_research_state)
